distsys/mpcal_context.go

```diff
@@ -0,0 +1,217 @@
+package distsys
+
+import (
+	"errors"
+	"fmt"
+	"github.com/benbjohnson/immutable"
+)
+
+var ErrAssertionFailed = errors.New("assertion failed")
+
+var ErrCriticalSectionAborted = errors.New("MPCal critical section aborted")
+
+// ArchetypeResourceHandle encapsulates a reference to an ArchetypeResource.
+// These handles insulate the end-user from worrying about the specifics of resource lifetimes, logging, and
+// crash recovery scenarios.
+type ArchetypeResourceHandle struct {
+	path TLAValue
+}
+
+// ArchetypeResourceMaker encapsulates how a specific kind of ArchetypeResource is created.
+// At its simplest, Make() should produce a fresh instance, and Configure will then be called on that instance
+// in order to do any further configuration to it.
+// This two-step process anticipates situations where the ArchetypeResource has been e.g reloaded from disk
+// during crash recovery, but might still need some configuration (setting up any function objects, user-provided Go channels).
+type ArchetypeResourceMaker interface {
+	Make() ArchetypeResource
+	Configure(res ArchetypeResource)
```

Comment:
I didn't understand why we need `Configure`.

Reply:
Can you explain what's missing from my attempt at explaining that precise issue in the comment?